url1a745d67fc_sobercheg_InterviewTraining_tgz-pJ8-leetcode_oj_Solution2TestJ8/src/careercup/linkedin/WordPatternMatcher.java

```diff
@@ -0,0 +1,58 @@
+package careercup.linkedin;
+
+import java.util.*;
+
+/**
+ * Created by Sobercheg on 12/18/13.
+ * http://www.careercup.com/question?id=12179920
+ * Given a large document and a short pattern consisting of a few words (eg. W1 W2 W3),
+ * find the shortest string that has all the words in any order (for eg. W2 foo bar dog W1 cat W3 -- is a valid pattern)
+ */
+public class WordPatternMatcher {
+
+    public String getShortestString(String input, List<String> patterns) {
+        // Idea: keep latest positions of every pattern word.
+        // If an input word matches a pattern word:
+        //  - update its position
+        //  - recalculate min distance as: currentPosition - min(pattern word positions)
+        //  - remember start and end positions
+
+        String[] inputTokens = input.split("[\\s]+");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;input&#39; (index: 0 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;patterns&#39;. The same parameter type is validated here for example: njr_failed/url1a745d67fc_sobercheg_InterviewTraining_tgz-pJ8-leetcode_oj_Solution2TestJ8/src/careercup/google/AlphaPalindrome.java:54. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).